src/WebJobs.Script/HostOptionsProvider.cs

```diff
@@ -0,0 +1,100 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+using Microsoft.Azure.WebJobs.Host.Config;
+using Microsoft.Azure.WebJobs.Host.Scale;
+using Microsoft.Azure.WebJobs.Hosting;
+using Microsoft.Azure.WebJobs.Script.Configuration;
+using Microsoft.Azure.WebJobs.Script.Extensions;
+using Microsoft.Extensions.Configuration;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+using Newtonsoft.Json;
+using Newtonsoft.Json.Linq;
+using Newtonsoft.Json.Serialization;
+
+namespace Microsoft.Azure.WebJobs.Script
+{
+    public class HostOptionsProvider : IHostOptionsProvider
+    {
+        private readonly JsonSerializer _serializer = JsonSerializer.Create(new JsonSerializerSettings
+        {
+            ContractResolver = new CamelCasePropertyNamesContractResolver(),
+        });
+
+        private readonly IEnumerable<IExtensionOptionsInfo> _extensionOptionsConfigurations;
+        private readonly IServiceProvider _serviceProvider;
+        private readonly ILogger<HostOptionsProvider> _logger;
+        private readonly IOptionsMonitor<ConcurrencyOptions> _concurrencyOptions;
+
+        public HostOptionsProvider(IServiceProvider provider, IEnumerable<IExtensionOptionsInfo> extensionOptionsConfigurations, IOptionsMonitor<ConcurrencyOptions> concurrencyOption, ILogger<HostOptionsProvider> logger)
+        {
+            _extensionOptionsConfigurations = extensionOptionsConfigurations;
+            _serviceProvider = provider;
+            _logger = logger;
+            _concurrencyOptions = concurrencyOption;
+        }
+
+        public string GetPayload()
+        {
+            var payload = new JObject();
+            var extensions = new JObject();
+            var extensionOptions = GetExtensionOptions();
+            foreach (var extension in extensionOptions)
+            {
+                extensions.Add(extension.Key, extension.Value);
+            }
+            if (extensionOptions.Count != 0)
+            {
+                payload.Add("extensions", extensions);
+            }
+
+            var concurrency = GetConcurrencyOptions();
+            payload.Add("concurrency", concurrency);
+            return payload.ToString();
+        }
+
+        private Dictionary<string, JObject> GetExtensionOptions()
+        {
+            var configuration = _serviceProvider.GetService<IConfiguration>();
+            // for each of the extension options types identified, create a bound instance
+            // and format to JObject
+            Dictionary<string, JObject> result = new Dictionary<string, JObject>();
+            foreach (IExtensionOptionsInfo extensionOptionsConfiguration in _extensionOptionsConfigurations)
```

Comment:
It should be the case that the collection only contains a single ExtensionOptionsInfo per options type, but I think theoretically if an extension did the wrong thing and called configure multiple times, we might end up adding multiple. So we should dedupe here for safety.

Reply:
Good point! Done!